cql/src/main/java/org/apache/cassandra/stargate/metrics/ClientMetrics.java

```diff
@@ -180,42 +249,95 @@ private int countConnectedClients() {
+    gauge.register(rows, true);
+  }
 
-    return clients;
+  /**
+   * Resolves a metric name for the micrometer registry.
+   *
+   * @param name metric name
+   * @return full name
+   */
+  private static String metric(String name) {
+    // to keep the back-ward compatibility, init all metric names with cql_ + DefaultNameFactory
+    String metricName = factory.createMetricName(name).getMetricName();
+    return "cql." + metricName;
   }
 
-  private List<Map<String, String>> recentClientStats() {
-    List<Map<String, String>> stats = new ArrayList<>();
+  private class ConnectionMetricsImpl implements ConnectionMetrics {
+
+    private final Tags tags;
+    private final Counter requestsProcessed;
+    private final Counter requestsDiscarded;
+    private final Counter authSuccess;
+    private final Counter authFailure;
+    private final Counter authError;
+
+    public ConnectionMetricsImpl(ClientInfo clientInfo) {
+      tags =
+          Optional.ofNullable(clientInfo)
+              .map(clientInfoTagProvider::getClientInfoTags)
+              .orElse(Tags.empty());
+
+      requestsProcessed = meterRegistry.counter(REQUESTS_PROCESSED_METRIC, tags);
+      requestsDiscarded = meterRegistry.counter(REQUESTS_DISCARDED_METRIC, tags);
+      authSuccess = meterRegistry.counter(AUTH_SUCCESS_METRIC, tags);
+      authFailure = meterRegistry.counter(AUTH_FAILURE_METRIC, tags);
+      authError = meterRegistry.counter(AUTH_ERROR_METRIC, tags);
+    }
+
+    @Override
+    public void markRequestProcessed() {
+      requestsProcessed.increment();
+    }
 
-    for (Server server : servers)
-      for (ClientStat stat : server.recentClientStats()) stats.add(stat.asMap());
+    @Override
+    public void markRequestDiscarded() {
+      requestsDiscarded.increment();
+    }
 
-    stats.sort(Comparator.comparing(map -> map.get(ClientStat.PROTOCOL_VERSION)));
+    @Override
+    public void markAuthSuccess() {
+      authSuccess.increment();
+    }
 
-    return stats;
-  }
+    @Override
+    public void markAuthFailure() {
+      authFailure.increment();
+    }
 
-  private <T> Gauge<T> registerGauge(String name, Gauge<T> gauge) {
-    return metricRegistry.register(factory.createMetricName(name).getMetricName(), gauge);
-  }
+    @Override
+    public void markAuthError() {
+      authError.increment();
+    }
 
-  private Histogram registerHistogram(String name) {
-    return metricRegistry.histogram(factory.createMetricName(name).getMetricName());
-  }
+    @Override
+    public Tags getTags() {
+      return tags;
+    }
 
-  private Counter registerCounter(String name) {
-    return metricRegistry.counter(factory.createMetricName(name).getMetricName());
-  }
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      ConnectionMetricsImpl that = (ConnectionMetricsImpl) o;
+      return tags.equals(that.tags);
```

Comment:
Would it be more human-readable to do all aggregations explicitly by `Tags` as opposed to by `ConnectionMetrics`, which delegates `equals` to `Tags`?

Reply:
Sure, done.